### ISSUE TITLE
Rustup (including proxies) is not safe for concurrent use

### ISSUE THREAD

**Issue by matklad** [rust-lang/rustup#988]
(Edited to capture all the details that have emerged over time)

# Recovering from this bug:

Usually just doing a `rustup component remove NAME && rustup component add NAME` will fix things. Sometimes removing the entire toolchain will be needed. In rare cases uninstalling rustup entirely will be needed.

# User model of Rustup

Rustup may be run as three different commands:
* rustup-init to install rustup (& by default a toolchain)
* rustup to explicitly query or modify an installation (including rustup itself and one or more toolchains)
* as a proxy of rustc, cargo etc (& which can implicitly trigger installation, upgrade of modification of a toolchain e.g. through toolchain files)

# Locking in Rustup

Rustup gets run concurrently in two very different contexts: within a single (machine, user), it may be run concurrently by the user, or an IDE, or both, to perform tasks ranging from toolchain installation, component addition, documentation opening. All of these require multiple reads to be made to the rustup data with a consistent view; some of them require writes to be made as well.

Rustup may also be run across machines, where a different machine but a shared rustup install is present - (machineA, userA) + (machineB, userA) - and in this case the same set of operations may take place, with the same requirements. 

Whatever consistency solution we adopt would be best if it can deliver both use cases, and not require manual lock administration as lockdir style solutions do, nor additional running network daemons.

## Proxies

Rustup has one set of proxies shared across all toolchains; the proxies are held open while any process from a toolchain is running - e.g. IDE's hold rls open for extended periods.

We need a lock to prevent concurrent attempts to install new proxies, and we need a notification mechanism back to the running proxies to allow them to be notified to exit when an update is required (because of presumed limitations of in-use-file-replacement on Windows, though recent changes may mean we can avoid this)

## Toolchains

We have many toolchains in one rustup installation; stable, beta, nightly, dated nightly versions, and custom versions. Adding a toolchain adds a directory and a hash file; we need a lock to prevent collisions attempting to move the directory into place. Deleting a toolchain does a recursive rm in-place, which also needs a lock to prevent other rustup invocations presuming that the toolchain is actually installed during the time the deletion takes place (or perhaps we need to rename-then-delete, though that can run into race conditions with virus scanners, especially if the toolchain was just installed). Further, permitting deletions at any point will require notifications to running rls process proxies from that toolchain to cause them to shutdown, or the .exe is likely not deletable on Windows.

## Components

Components such as `rls` are added into a toolchain directory, and also involve writing to a metadata file within the toolchain tree itself. This needs to be locked to avoid corruption/dropped writes. As with toolchains, we need proxy notification for component removal, as well as a way to make sure that a component that is being removed does not have new instances of it spawned between the start of the removal and the completion of the removal.

## Downloads

We download packages for rustup itself, toolchains and additional components for toolchains, and (a corner case) custom installer executables for toolchains. We also download digital signature metadata files.

The same file can be downloaded by two different rustup invocations for two different reasons. For instance, downloading nightly and a dated nightly for today, will download the same file(s).

We used to leak partially downloaded files, and recently started deleting all download dir contents when rustup finished running; this is causing errors now.

We need some mechanism to deal with leaks, but also to permit concurrent execution of rustup to be downloading files without interruption. Possibly a date based mechanism or locking based mechanism would be sufficient.

## Network file systems & rustup

Linux handles SMB mounts and locking on that filesystem well, at least per my reading of the source - a rustup dir on an SMB mounted file system using regular posix locks should have those locks reflected as SMB locks.

NFS is well known for having poor lock behaviour when the network services are not running or are firewalled; the underlying syscalls are only non-blocking on the filedescriptor themselves, and EWOULDBLOCK is defined as the lock being already held, not the OS being unable to determine if the lock is already held...
```
  [EWOULDBLOCK]
	    The file is locked and the LOCK_NB option was specified.
```
So it is likely that we will see bug reports of rustup hanging when `~/.rustup` is on NFS and the NFS server's lock RPC service is misconfigured or flaky. 
I think this be mitigated by emitting an NFS specific log message when taking a lock out on NFS once per process; with a config option to disable that for users that are confident they don't need it.... and a bug reporting hint to tell us they have disabled it.

## Locks and notifications

OS locks typically don't have callback notifications built in; polling is one answer, or some form of lightweight message bus (e.g. zmq) with clear rules tied into the lock management layer. We have to be careful about race conditions though: in particular notifying before or after as appropriate.

**Comment by brson** [comment]
rustup needs to have concurrency support bolted on in general. Need to extract cargo's flocking code.

**Comment by brson** [comment]
To start with we can just bolt on a flock to the entire runtime of rustup.

**Comment by matklad** [comment]
I think I can probably work on this: I have some experience with Cargo's file [dead]locking :)

**Comment by brson** [comment]
@matklad ooh awesome. Can you extract cargo's fancy flock code to its own crate to share with rustup?

**Comment by brson** [comment]
Oh there's a tricky bit here in that rustup is reentrant. Can't just use a simple global lock.

**Comment by matklad** [comment]
>Oh there's a tricky bit here in that rustup is reentrant. Can't just use a simple global lock.

Where exactly does reentrancy happen? 

I don't yet understand how rustup works, but looks like adding a file lock to the `Transaction` should do the trick, unless transactions can nest and deadlock each other?

**Comment by fenhl** [comment]
Hi, is this still an issue? It might explain some problems I've been having with rustup.

**Comment by kinnison** [comment]
Yes, currently we have no locking on the toolchain directories IIRC.

**Comment by rbtcollins** [comment]
@matklad did that fancy flock code get extracted?

**Comment by matklad** [comment]
No: https://github.com/rust-lang/cargo/blob/master/src/cargo/util/flock.rs

I’ve also absolutely dropped the ball on this one :)

**Comment by rbtcollins** [comment]
No worries. Looking at it the semantics you have are different enough that it wouldn't be useful (e.g. creating directories on demand would run open up TOCTOU race conditions in untar secure handling for rustup), we don't want to lock each file because syscall latency would destroy performance on NFS and windows and so on.

I'll take the code as inspiration and do a similar thing though :) cheers!

**Comment by rbtcollins** [comment]
Ok, so looks like cargo has problems on NFS which I was worried about: 
- https://github.com/rust-lang/cargo/blob/master/src/cargo/util/flock.rs#L300 
- https://github.com/rust-lang/cargo/issues/4629

cifsfs supports oplocks so SMBfs should lock properly: 
- https://github.com/torvalds/linux/blob/master/fs/cifs/cifsfs.c
- meaning that I think its just the blocking failure mode of crashed /firewalled/missing rpc.lockd that we need to be concerned with. 

fs2 which cargo uses does use LOCK_NB with flock - flock(file, libc::LOCK_EX | libc::LOCK_NB) (https://tantivy-search.github.io/tantivy/src/fs2/unix.rs.html#36)

questions: 
- would lockf instead, with F_TLOCK still fail in the same way? It is defined as never blocking, but network file systems ... need to check the code.
- Open file description locks still don't have a timeout field in the lock descriptor that calls operate on, so there's no way to tell NFS that you don't want to wait for that shitty shitty network. There is ENOLCK to signal that locking over the network *failed*, but the question is how long the network takes to realise it.
- https://docs.oracle.com/cd/E19253-01/816-4555/rfsrefer-9/index.html has a client side retransmit timer of 15 seconds, which suggests ELONGTIME.

I'm inclined to provide a UI warning about NFS when taking the lock with a env variable to shut the warning up for folk that know things work properly in their environment (using the NFS sniffing logic from cargo); but then assume that the fs is working properly.

**Comment by rbtcollins** [comment]
Further thoughts: we need to mutually exclude operations on toolchains: removal and upgrade affect the entire toolchain as well as operations on components (add/remove components). I think locking the dir of the toolchain is probably the right control point.

**Comment by metajack** [comment]
We've been running into this on the Libra team as developers using CLion or IDEA manage to invoke rustup concurrently accidentally quite often when we check in toolchain file updates. This results in a borked toolchain install that is missing rustc, but fixes itself after a manual run of `rustup toolchain install $TOOLCHAIN`.

@rbtcollins Are you working on a PR to fix this now? If so, that would be awesome.

**Comment by rbtcollins** [comment]
Not actively working on a PR, but it is on my radar.

**Comment by rbtcollins** [comment]
Ok, so here's a bit of a specification, I think this ties together all the various bits involved.

# Locking in Rustup

Rustup gets run concurrently in two very different contexts: within a single (machine, user), it may be run concurrently by the user, or an IDE, or both, to perform tasks ranging from toolchain installation, component addition, documentation opening. All of these require multiple reads to be made to the rustup data with a consistent view; some of them require writes to be made as well.

Rustup may also be run across machines, where a different machine but a shared rustup install is present - (machineA, userA) + (machineB, userA) - and in this case the same set of operations may take place, with the same requirements. 

Whatever consistency solution we adopt would be best if it can deliver both use cases, and not require manual lock administration as lockdir style solutions do, nor additional running network daemons.

## Proxies

Rustup has one set of proxies shared across all toolchains; the proxies are held open while any process from a toolchain is running - e.g. IDE's hold rls open for extended periods.

We need a lock to prevent concurrent attempts to install new proxies, and we need a notification mechanism back to the running proxies to allow them to be notified to exit when an update is required (because of presumed limitations of in-use-file-replacement on Windows, though recent changes may mean we can avoid this)


## Toolchains

We have many toolchains in one rustup installation; stable, beta, nightly, dated nightly versions, and custom versions. Adding a toolchain adds a directory and a hash file; we need a lock to prevent collisions attempting to move the directory into place. Deleting a toolchain does a recursive rm in-place, which also needs a lock to prevent other rustup invocations presuming that the toolchain is actually installed during the time the deletion takes place (or perhaps we need to rename-then-delete, though that can run into race conditions with virus scanners, especially if the toolchain was just installed). Further, permitting deletions at any point will require notifications to running rls process proxies from that toolchain to cause them to shutdown, or the .exe is likely not deletable on Windows.

## Components

Components such as `rls` are added into a toolchain directory, and also involve writing to a metadata file within the toolchain tree itself. This needs to be locked to avoid corruption/dropped writes. As with toolchains, we need proxy notification for component removal, as well as a way to make sure that a component that is being removed does not have new instances of it spawned between the start of the removal and the completion of the removal.

## Downloads

We download packages for rustup itself, toolchains and additional components for toolchains, and (a corner case) custom installer executables for toolchains. We also download digital signature metadata files.

The same file can be downloaded by two different rustup invocations for two different reasons. For instance, downloading nightly and a dated nightly for today, will download the same file(s).

We used to leak partially downloaded files, and recently started deleting all download dir contents when rustup finished running; this is causing errors now.

We need some mechanism to deal with leaks, but also to permit concurrent execution of rustup to be downloading files without interruption. Possibly a date based mechanism or locking based mechanism would be sufficient.

## Network file systems & rustup

Linux handles SMB mounts and locking on that filesystem well, at least per my reading of the source - a rustup dir on an SMB mounted file system using regular posix locks should have those locks reflected as SMB locks.

NFS is well known for having poor lock behaviour when the network services are not running or are firewalled; the underlying syscalls are only non-blocking on the filedescriptor themselves, and EWOULDBLOCK is defined as the lock being already held, not the OS being unable to determine if the lock is already held...
```
  [EWOULDBLOCK]
	    The file is locked and the LOCK_NB option was specified.
```
So it is likely that we will see bug reports of rustup hanging when `~/.rustup` is on NFS and the NFS server's lock RPC service is misconfigured or flaky. 
I think this be mitigated by emitting an NFS specific log message when taking a lock out on NFS once per process; with a config option to disable that for users that are confident they don't need it.... and a bug reporting hint to tell us they have disabled it.

## Locks and notifications

OS locks typically don't have callback notifications built in; polling is one answer, or some form of lightweight message bus (e.g. zmq) with clear rules tied into the lock management layer. We have to be careful about race conditions though: in particular notifying before or after as appropriate.

**Comment by kinnison** [comment]
One aspect of `rustup`'s data which needs locking and isn't on that list is the configuration (`settings.toml`) which can be altered by a number of `rustup` commands such as `rustup default XXX` or `rustup override ...`

**Comment by ckaran** [comment]
Is this still an issue?  What about doing all work in the system's temp directory, and atomically renaming into the right directory afterwards?

There are two reasons I'm thinking about this solution. First, the system should handle cleaning up the temporary directory automatically.  That means that if `rustup` crashes or does something else weird dribbling files around in the equivalent of `/tmp`, they will all be deleted by the system at some point in the future.

Second, IIRC file and directory renames are atomic in POSIX-compliant filesystems (maybe others as well?  This needs to be verified).  So, much like doing a compare and swap operation on an atomic pointer, only the first process that succeeds has any effect.  Other processes/threads will get an error back saying that they failed the rename.  This deals with many of the issues that locking is intended to solve.

All that said, replacing a toolchain may be more difficult because (IIRC) if a process is executing renaming the path to the executable has no effect on it.  This will lead to a TOCTOU where a user starts using an old executable, the new one is atomically swapped in to replace the old one, and the old one looks for resources within the new directory structure.  E.g., someone does `rustup doc` and `rustup update` concurrently, with the switching between processes being just awful enough that the old version of `rustup doc` is ends up searching within the new directory structure.  I have no good answer for this.

**Comment by rbtcollins** [comment]
Yes it is. We already use atomic operations to the extent the filesystem in
use can deliver them, but we don't build a lock out of them. Without a lock
the range operations we perform lead to inconsistencies, further to that
there is a performance issue and as you say, changing state underneath
running processes is unkind.

**Comment by ckaran** [comment]
@rbtcollins OK.  Just to clarify my own understanding of the issue as @matklad described it, if the update crashes while it is happening, it could leave the system in an indeterminate state, which is enough of a problem that we want some method of avoiding it completely, correct?  If that is true, below is a first stab at solving the two problems.  Every process is required to follow the steps below, or everything fails.

1. Create a new temporary file in `/tmp`.  My current understanding is that this is not guaranteed to be atomic, but that doesn't matter, the only requirement is that they try to create this file.
2. Each process then tries to atomically rename their own temporary file to some well-known name, e.g. `/tmp/rustup_lock_file`.  Bonus points if the contents of the file include a short readme so that anyone wondering what this file is can get an idea as to what it's for.  It may even contain a URL pointing to this issue, or something similar.
3. If the rename fails, it means that either some other running process owns the lock, or the other process crashed while holding the lock.  In most cases it will mean the former, and as long as everything is proceeding as expected nobody will look at this file, and it will either be cleared at the end of the second phase (see below), or on the next reboot of the machine which is when the system will reset the contents of `/tmp`.
4. If the rename succeeds, then that process now owns the lock.  It now starts the second stage actions.
     1. It looks for a file within `~/.rustup` named `state_machine`.  If this file does not exist, then the last process to do any work succeeded, and the system is in a clean state.  Since the process owns the lock, it can now safely create the `state_machine` file, and update its contents with status messages, state information, etc., etc., etc. safely.  Once all work is done, the process first deletes the `state_machine` file, and then deletes the `/tmp/rustup_lock_file`, confirming that the former is truly committed before executing the latter.
     2. If the file does exist, it means that the last process to be doing updates failed somehow, leaving the directory in an unclean state.  The state written within the file could be used to indicate what steps succeeded, and what steps need to be rolled back or finished, or whatever.  Regardless, cleanup steps are initiated.

There are a few advantages to the above algorithm:
- There is only one lock, so the time it takes to grab or release the lock should be small relative to the amount of work the rest of the system is doing.
- It doesn't preclude the use of threads internal to a process, so we're not forcing single threading, just single processing.
- If the process crashes, then cleaning up is simply a matter of rebooting the system.  Since that seems to be the go-to fix for everything computer related, it won't be a surprise to anyone to find out that's what they should do.  The advantage is that if the process crashed due to something weird in the computer itself unrelated to `rustup`, rebooting might clear that issue out, letting the process continue to completion after reboot.
- The `state_machine` file gives a hint as to where the process was when it failed.  This could be useful when debugging what's going on, and for general awareness of the current state of the process (e.g., `tail -f ~/.rustup/state_machine` is a quick and dirty way of watching the current progress of the process).

The one big disadvantage is that this is **not** a truly transactional algorithm.  I don't yet know how to do that, but maybe some of the database literature would help with that?

**Comment by Xuanwo** [comment]
For those encountering a similar problem or directed to this issue, here's a quick workaround from https://github.com/rust-lang/rustup/issues/3530:

Ensure the rust-toolchain is set up by running `cargo version` before using other tools like `matruin`.

**Comment by duaneking** [comment]
Just found this issue again and the solution seemed to be to close all Visual Studio Code instances before doing the ``rustup update`` and that worked, but unfortunately it does look like Visual Studio Code is locking these files despite the windows not being active or in use.

**Comment by rami3l** [comment]
After thinking about this issue for quite a while, I have formed the following early ideas around a hypothetical file-based global RwLock:

- "Rustup mode" (i.e. the `rustup` command, called "Rustup" below): Acquires write lock on enter, releases on exit.
- "Proxy mode" (i.e. the shims such as `cargo` and `rust-analyzer`, called "Proxy" below): Acquires read lock on enter, releases on exit.
  - Downside: we might no longer be able to use `exec` on Unix, unless we can get the cleanup code running when the current PID is repurposed (which seems possible by using `flock` on dummy files).

The annoying thing here is that we also have to consider the subprocesses we're creating (since the critical sections could be nested), so:

- Rustup calling Rustup: IIRC this happens in tests. If we do this carefully enough, we should be able to handle it correctly... Possibly by making the lock re-entrant.
- Rustup calling proxy: I don't believe this is happening anywhere in the codebase...
- Proxy calling Rustup: The subprocess should somehow upgrade the parent's read lock to a write lock.
- Proxy calling proxy: The subprocess should somehow inherit the read lock from the parent.

... and it looks like if we're making a re-entrant lock, the uniqueness of its ownership should be based on the root of the process' subtree (up to a certain Rustup/proxy). This info might be conveniently passed on via a new env variable and/or via the lock file itself. Something like [C#'s "named mutex"](https://learn.microsoft.com/en-us/dotnet/api/system.threading.mutex.-ctor?view=net-8.0#system-threading-mutex-ctor(system-boolean-system-string-system-boolean@)) is close to what I'm looking for, which happens to be implemented using lock files on Unix.

All these points rely on the fact that a proxy cannot implicitly modify the current installation (so can be safely granted read-only access), which has fortunately been ensured by #3635.

Further down the line, more "Rustup mode" operations could be made to acquire read locks by default instead, and these locks might better be managed on a per-toolchain basis.

I'm actively trying to unblock this issue, although I'm not sure about the correctness of this reasoning... I'm trying this out in a local repo... Any input is welcome. 🤔 

---

Update: @weihanglo suggests that such a RwLock already exists in Cargo (or more precisely, its variation, as it holds [two](https://github.com/rust-lang/cargo/blob/9736f3204a2599d6a152d9bd3d800f5a6124d175/src/cargo/util/cache_lock.rs#L344-L358) [reentrant file-based](https://github.com/rust-lang/cargo/blob/9736f3204a2599d6a152d9bd3d800f5a6124d175/src/cargo/util/cache_lock.rs#L160-L174) locks to form a custom [3-mode](https://github.com/rust-lang/cargo/blob/9736f3204a2599d6a152d9bd3d800f5a6124d175/src/cargo/util/cache_lock.rs#L97-L137) global RwLock), so that should be a solid starting point.

cc @rbtcollins what was the concern in https://github.com/rust-lang/rustup/issues/988#issuecomment-582285054 that made you think Cargo's lock has semantics different enough from what we'd expect?

**Comment by rami3l** [comment]
Regarding https://github.com/rust-lang/rustup/issues/988#issuecomment-2340369437, I've made a working on Unix PoC for a mocked `rustup`/`cargo`/`rust-analyzer` scenario using a combination of `execvp()`, `fcntl()` and `open()` syscalls on a marker file.

The outcome has been positive overall: `rustup` commands, for example, when trying to acquire a write lock, would be blocked by `rust-analyzer`'s read lock.

The two key points mentioned in the previous comment seem to have been implemented, namely the inheritance of locks from parent process to child, and the upgrade of a read lock to a write lock on `rustup` commands, thanks to the fact that `open()` works across `exec*()` if `CLOEXEC` is not set (note: this is unfortunately not the case with `std::fs`).
The currently-open FD is passed to the subprocess via an extra env variable, where its ownership is recovered by `File::from_raw_fd()`. This seems alright in terms of handle ownership, but I'm not quite sure if this is acceptable security-wise.

However, this comes with a notable limitation: since these properties (lock inheritance and lock upgrade) are limited to subprocesses launched using `exec*()`, this implementation might deadlock any proxy trying to run `rustup` commands (in incompatible ways) that require a write lock, e.g. `cargo run` running a program that uses `Command::spawn()` to run `rustup toolchain install`.
Fortunately, thanks to the shared nature of the read lock, other use cases (including proxies) seem to work without issues, regardless of the nested critical sections.

The lock is currently not reentrant in-process, but when integrating the solution into `rustup`, it should be relatively easy to adapt `cargo`'s approach to make it so.

Next up, I'll try to clean up the PoC and make it public for reference.

There are still many questions to be answered, to name a few:
- Should the aforementioned limitation be considered tolerable?
  - If so, what can we do to better communicate the contract with devtool authors (that `rustup` should be executed with `exec*()` only)?
  - How many users/repos are currently launching `rustup` through `Command::spawn()` and friends? How should we help handle the potential migration?
  - NB: `cargo` suffers from the same limitation: https://github.com/rust-lang/cargo/issues/6412
  - If not, how can we improve this locking scheme?
- How can we get something similar working on Windows?

**Comment by ChrisDenton** [comment]
I think Windows is a bit simpler here. On Windows the rustup process stays alive for the duration of the proxied binary no?

Locking a file can be done either by setting the [`share_mode`](https://doc.rust-lang.org/std/os/windows/fs/trait.OpenOptionsExt.html#tymethod.share_mode) or by using [`LockFileEx`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex).

A third option would be to craft an rwlock from [interprocess synchronization](https://learn.microsoft.com/en-us/windows/win32/sync/interprocess-synchronization) objects. However this may be trickier to get right.

**Comment by rami3l** [comment]
> I think Windows is a bit simpler here. On Windows the rustup process stays alive for the duration of the proxied binary no?

@ChrisDenton That's where I got stuck.

It seems to me that the previous Unix example where someone uses `cargo run` to run `rustup toolchain install` with `Command::spawn()` also applies to Windows.

In this case, `cargo` launches `rustup` as a subprocess, and the two processes coexist. `cargo` acquires the read lock first, so `rustup` has no simple way to get a write lock unless it somehow has a way to inform `cargo` to abandon the lock temporarily when the read lock is in place. 

OTOH, if on Unix we use `Command::exec()` (calling `execvp()` underneath), `rustup` will *replace* `cargo`, inheriting the latter's read lock, which can be easily upgraded to a write lock later on. On Windows, such an escape hatch doesn't exist, meaning proxies have no good ways of calling `rustup`.

Of course, we can admit the limitation and move on (maybe by supporting something like `RUSTUP_UNSAFE_NO_LOCK=1`), but whether we should do this remains debatable. I'm just pointing this thing out.

**Comment by ChrisDenton** [comment]
The flow is something like `rustup -> cargo -> rustup`, right? So there just needs to be a way to communicate back to the "root" `rustup` (and indeed to know if an ancestor rustup exists in the first place).

Though an issue I can see is that `LockFileEx` doesn't support upgrading locks. You'd have to release and reacquire, which would allow a window of opportunity for someone else to grab a write lock. Maybe that can be mitigated though.

**Comment by rami3l** [comment]
Oh wait, there's an [`execvp()`](https://docs.rs/libc/0.2.158/x86_64-pc-windows-msvc/libc/fn.execvp.html) function in Windows CRT, although I believe what it truly does is making the parent process quit immediately after launching the subprocess. It does help with unlocking though. I'm not sure if it fits our use case here.

If it does, I still think my approach would be acceptable regardless of the current limitation, especially when the RwLock becomes per-toolchain (when `rustup` is called to modify the current toolchain despite the fact that `cargo` is running on the same toolchain, something's gone clearly wrong).

**Comment by ChrisDenton** [comment]
> Oh wait, there's an [`execvp()`](https://docs.rs/libc/0.2.158/x86_64-pc-windows-msvc/libc/fn.execvp.html) function in Windows CRT, although I believe what it truly does is making the parent process quit immediately after launching the subprocess. It does help with unlocking though.

Yes, though that would break anything that waits on rustup process. Such as the console.

> If it does, I still think my approach would be acceptable regardless of the current limitation, especially when the RwLock becomes per-toolchain (when `rustup` is called to modify the current toolchain despite the fact that `cargo` is running on the same toolchain, something's gone clearly wrong).

Fair enough! 👍

**Comment by rami3l** [comment]
> > Oh wait, there's an [`execvp()`](https://docs.rs/libc/0.2.158/x86_64-pc-windows-msvc/libc/fn.execvp.html) function in Windows CRT, although I believe what it truly does is making the parent process quit immediately after launching the subprocess. It does help with unlocking though.
> 
> Yes, though that would break anything that waits on rustup process. Such as the console.

@ChrisDenton Thanks for the info.

Given that, I guess it might make some sense to go back and redesign a lock manager daemon of some sort, as suggested by @rbtcollins, instead of communicating with the parent Rustup instance directly:

> my thinking - not well written up - is that a persistent server process would make this locking a lot easier for a local workstation, at the cost of not working on a network drive. There are fancy mechanisms to do locks with just atomic fs operations but really, in 2024 networked setups are rarely developer environments, much more likely to be buildfarms where a single head node is all that needs to be configured.
>
> A persistent server process could (for instance) start implicitly per-toolchain, shutdown after 10 seconds of inactivity, and while running broker locks, logs, interact with vscode etc. But its a bit of a bigger concept and I know some folk are highly allergic to the very idea. (Though I bet they still use a web browser and other tools that have a similar background task)
_https://discord.com/channels/442252698964721669/463480252723888159/1285284728341266452_

This also helps ensuring cross-platform compatibility without having to deal with platform specifics like I currently do.

Also, I think this should play well with current self-updating mechanisms, as long as the updater waits until both the locker instance and the existing `rustup` instances to terminate.

**Comment by cratelyn** [comment]
i got bit by this issue today. i was able to solve this by stopping running rust-analyzer processes, uninstalling my rustup toolchains, and trying again. the error message provided was not particularly helpful or clear though, by the rust project's usual standards.

thank you for the remediation steps outlined here.

**Comment by georglauterbach** [comment]
We have issues being closed as duplicates of this issue, but nothing is moving here. I am constancly experiencing this issue, and I cannot explain where the concurrency is coming from - I am certainly not executing `rustup` concurrently.

While things work locally, it does not in our CI environment - but for no apparent reason. Is there anything we can do to mitigate this bug in GitLab's CI?

---

Moreover, this issue is becoming **10 years old** soon. Who can tackle it? What is generally required?

I am willing to spend time on this, but I need support. I think I can provide infrastructure so we can talk this through via a meeting, which by now is probably necessary, because this thread is already very long.

I have seen some folks have already worked on this. @rbtcollins @matklad @rami3l Would you be open to tackling this together?

**Comment by djc** [comment]
> Moreover, this issue is becoming 10 years old soon. Who can tackle it? What is generally required?

We're all volunteers here. Do you want to fund work on this?

**Comment by georglauterbach** [comment]
> > Moreover, this issue is becoming 10 years old soon. Who can tackle it? What is generally required?
> 
> We're all volunteers here. Do you want to fund work on this?

I have no funds to spend on this, and I know what it is like being a volunteer on a FOSS project. I think it's more important to have developers spend time on this instead of donating money - I am willing to do that to help. But I cannot do it alone, I do not have the time to tackle this whole thing myself. But I can bring those who can spend time on this together and then we can work together.

**Comment by djc** [comment]
> I have no funds to spend on this, and I know what it is like being a volunteer on a FOSS project. I think it's more important to have developers spend time on this instead of donating money - I am willing to do that to help. But I cannot do it alone, I do not have the time to tackle this whole thing myself. But I can bring those who can spend time on this together and then we can work together.

Unfortunately this is (as far as I know) one of those things that requires both fairly intimate knowledge of the inner workings of rustup as well as a fair bit of knowledge of different platforms (including Windows, macOS and Linux) to make it work. I think @rami3l has been working on it a little, but getting it done is quite a bit of work and not necessarily accessible to "casual" contributors. I'll let @rami3l comment on whether there's some work that can be peeled off here.

**Comment by Alexendoo** [comment]
For those running into this due to `rust-analyzer` running concurrently with a manual `cargo` invocation there's a nice workaround that's now possible:

```json
    "rust-analyzer.cargo.extraEnv": {
        "RUSTUP_AUTO_INSTALL": "0"
    },
```

RA will now error out instead of implicitly installing a toolchain. In a project with a `rust-toolchain.toml` you no longer have to think about waiting to run `cargo`/disabling RA when switching commits

**Comment by rbtcollins** [comment]
Setting the auto-install to false by default is a breaking change, but if we can swing the political discussion around it, it is the simplest and most effective way of fixing this issue.

We know people use rustup on NFS and SMB shared file systems, and while its not the dark ages of the 2000's anymore, I rather suspect we'll have a rich source of bugs over time from an insufficiently careful locking strategy, especially as there will be a (fairly predictable, sadly), tension between granularity and perceived performance.

Rustup wasn't built to be run concurrently, and Rust Analyzer (and other IDEs) violate that expectation.

Possibly also the RA team could set that variable automatically, as it is never safe for rust-analyzer to trigger auto-installs...

**Comment by ChrisDenton** [comment]
> Possibly also the RA team could set that variable automatically, as it is never safe for rust-analyzer to trigger auto-installs...

I made a ra issue to track this. See the issue linked above,

**Comment by rami3l** [comment]
> I think [@rami3l](https://github.com/rami3l) has been working on it a little, but getting it done is quite a bit of work and not necessarily accessible to "casual" contributors. I'll let [@rami3l](https://github.com/rami3l) comment on whether there's some work that can be peeled off here.

@djc Thanks for the cue! This has been one of my most-wanted since my joining the team and indeed I've done some preliminary research to settle the locking scheme, without which we cannot really start the implementation. Given the current roadmap I do plan to work on this one in the following order:

- Settle the GSoC 2025 project via #4471.
- Implement #3940 and replace the transaction system with it (I remember the previous team talked about removing the transaction system but I can't find the link for now).
- Decide on a locking scheme and try to work on a PoC.

When the time is right I'll post this as a half-year project goal to increase visibility and to request for comments.

My current conclusion is that @rbtcollins 's analysis was right. If we want to remain compatible with so many cases/platforms while ensuring correctness especially with recursive calls (this makes `rustup` fundamentally different from say, `apt` and `cargo`) , we might have to sacrifice some properties such as read-write locking. I still want to implement the "waiting for notification" aspect of this custom lock if at all possible, but I'm not really sure if OS-native mechanisms can do that job well.

**Comment by georglauterbach** [comment]
@rami3l that sounds very nice! :)

**Comment by konstin** [comment]
I accidentally added some context in a new issue at https://github.com/rust-lang/rustup/issues/4607 instead of here cause I didn't find the issue. As summary, I created a PR for fixing one of the parallelism issues using an exclusive file lock (https://github.com/rust-lang/rustup/pull/4606). This is the first issue I ran into when running `rustup-init -y` as build bootstrapping step in parallel. This lock has a small scope, only around the download and rename operation.

**Comment by jokeyrhyme** [comment]
i recently came across the file-`inotify()`-based protocol used by systemd for some IPC, which may be part of an approach for orchestrating mutual exclusion between processes: https://systemd.io/PASSWORD_AGENTS/

that said, it seems you already have a similar approach, and there's probably also a crate that covers this 🤷

**Comment by rami3l** [comment]
For anyone who is still bothered by this issue, I have successfully pushed for establishing its resolution as a proper project goal: https://github.com/rust-lang/rust-project-goals/pull/731

---

Update: The project goal's process is currently being tracked at:
- https://github.com/rust-lang/goals/issues/736
